cmd/src/actions_exec_logger.go

```diff
@@ -227,9 +230,126 @@ func (a *actionLogger) DockerStepStarted(repoName string, step int, image string
 }
 
 func (a *actionLogger) DockerStepErrored(repoName string, step int, err error, elapsed time.Duration) {
+	a.progress.IncStepsComplete(1)
+	a.progress.IncStepsFailed()
 	a.write(repoName, boldRed, "%s %s. (%s)\n", boldBlack.Sprintf("[Step %d]", step), err, elapsed)
 }
 
 func (a *actionLogger) DockerStepDone(repoName string, step int, elapsed time.Duration) {
+	a.progress.IncStepsComplete(1)
 	a.write(repoName, yellow, "%s Done. (%s)\n", boldBlack.Sprintf("[Step %d]", step), elapsed)
 }
+
+// write writes to RepoWriter and if in verbose mode, to stderr
+func (a *actionLogger) write(repoName string, c *color.Color, format string, args ...interface{}) {
+	if w, ok := a.RepoWriter(repoName); ok {
+		fmt.Fprintf(w, format, args...)
+	}
+	a.log(repoName, c, format, args...)
+}
+
+// log logs only to stderr, it does not log to our repoWriters
+func (a *actionLogger) log(repoName string, c *color.Color, format string, args ...interface{}) {
+	if !a.verbose {
+		return
+	}
+	if len(repoName) > 0 {
+		format = fmt.Sprintf("%s -> %s", c.Sprint(repoName), format)
+	}
+	fmt.Fprintf(a.progressWriter, format, args...)
+}
+
+type progress struct {
+	patchCount int64
+
+	totalSteps    int64
+	stepsComplete int64
+	stepsFailed   int64
+}
+
+type progressWriter struct {
+	p *progress
+
+	mu                sync.Mutex
+	w                 io.Writer
+	shouldClear       bool
+	progressLogLength int
+}
+
+func (w *progressWriter) Write(data []byte) (int, error) {
+	w.mu.Lock()
+	defer w.mu.Unlock()
+
+	if w.shouldClear {
+		// Clear current progress
+		fmt.Fprintf(w.w, "\r")
+		fmt.Fprintf(w.w, strings.Repeat(" ", w.progressLogLength))
+		fmt.Fprintf(w.w, "\r")
+	}
+
+	if w.p.TotalSteps() == 0 {
+		// Don't display bar until we know number of steps
+		w.shouldClear = false
+		return w.w.Write(data)
+	}
+
+	if !bytes.HasSuffix(data, []byte("\n")) {
+		w.shouldClear = false
+		return w.w.Write(data)
+	}
+
+	n, err := w.w.Write(data)
+	if err != nil {
+		return n, err
+	}
+	total := w.p.TotalSteps()
+	done := w.p.StepsComplete()
+	var pctDone float64
+	if total > 0 {
+		pctDone = float64(done) / float64(total)
+	}
+
+	maxLength := 50
+	bar := strings.Repeat("=", int(float64(maxLength)*pctDone))
+	if len(bar) < maxLength {
+		bar += ">"
+	}
+	bar += strings.Repeat(" ", maxLength-len(bar))
+	progessText := fmt.Sprintf("[%s] steps(%d/%d) %s %s", bar, w.p.StepsComplete(), w.p.TotalSteps(), boldRed.Sprintf("failed(%d)", w.p.TotalStepsFailed()), hiGreen.Sprintf("patches(%d)", w.p.PatchCount()))
+	fmt.Fprintf(w.w, progessText)
+	w.shouldClear = true
+	w.progressLogLength = len(progessText)
+	return n, err
+}
+
+func (p *progress) SetTotalSteps(n int64) {
```

Comment:
Nitpicky nitpick: I'd move the method definitions on `*progress` up to the definition of `*progress`.

Reply:
Ah, yeah, missed that after splitting it up